### PR TITLE
C++: support for fold expressions

### PR DESCRIPTION
**Pull request by nickrolfe**

### CODE REVIEW

**Comment by dave-bartolomeo** [comment]
When a template is instantiated, any fold expressions will just show up as the expanded sequence of plain expressions, right? `FoldExpr` will only appear in uninstantiated templates?

File: cpp/ql/src/semmle/code/cpp/exprs/Expr.qll [hunk 1]
```diff
@@ -964,6 +964,65 @@ class NoExceptExpr extends Expr, @noexceptexpr {
   }
 }
 
+/**
+ * A C++17 fold expression.
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/exprs/Expr.qll]: Please clarify that this expression only appears in templates, not in their instantiations.

**Comment by nickrolfe** [comment]
Yes, that's right. I could add another query to `select any(Expr e)` in the test, if you think that would be useful.

File: cpp/ql/src/semmle/code/cpp/exprs/Expr.qll [hunk 2]
```diff
@@ -964,6 +964,65 @@ class NoExceptExpr extends Expr, @noexceptexpr {
...
+   */
+  Expr getInitExpr() {
+    this.isBinaryFold() and
+    if this.isRightFold() then result = getChild(1) else result = getChild(0)
+  }
+
+  /**
+   * Holds if this is a binary fold with a child expression representing the initial value.
+   */
+  predicate hasInitExpr() { exists(this.getInitExpr()) }
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/exprs/Expr.qll]: I don't see a need for this predicate, other than to make users wonder whether it's the same as `isBinaryFold`.

File: cpp/ql/src/semmle/code/cpp/exprs/Expr.qll (hunk 1 above)
Comment (nickrolfe) [reply cpp/ql/src/semmle/code/cpp/exprs/Expr.qll]: Done.

File: cpp/ql/src/semmle/code/cpp/exprs/Expr.qll (hunk 2 above)
Comment (nickrolfe) [reply cpp/ql/src/semmle/code/cpp/exprs/Expr.qll]: I've deleted it.

**Pull request merged**